File: C++/2-Medium/00159._至多包含两个不同字符的最长子串/00159._至多包含两个不同字符的最长子串.cpp
#include <iostream>
#include <unordered_map>
#include <string>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstringTwoDistinct(string s) {
        unordered_map<char, int> freq; // 字符出现频率表
        int maxLen = 0; // 最长子串的长度
        int left = 0; // 滑动窗口左边界
        for (int right = 0; right < s.length(); ++right) {
            freq[s[right]]++; // 将右边界字符加入频率表

            // 如果窗口中的不同字符超过2个，则移动左边界直到不同字符数为2
            while (freq.size() > 2) {
                freq[s[left]]--; // 左边界字符出现频率减1
                if (freq[s[left]] == 0) {
                    freq.erase(s[left]); // 如果字符出现频率为0，则从表中删除该字符
                }
                left++; // 移动左边界
            }

            // 更新最长子串长度
            maxLen = max(maxLen, right - left + 1);
        }
        return maxLen;
    }
};

int main() {
    Solution solution;
    string s = "eceba";
    cout << "The length of the longest substring with at most two distinct characters is: "
         << solution.lengthOfLongestSubstringTwoDistinct(s) << endl;
    return 0;
}

File: Python/2-Medium/00159._至多包含两个不同字符的最长子串/00159._至多包含两个不同字符的最长子串.py
from typing import Dict

def length_of_longest_substring_two_distinct(s: str) -> int:
    """
    寻找最多包含两个不同字符的最长子串的长度。

    Args:
    s (str): 输入字符串。

    Returns:
    int: 最长子串的长度。

    示例:
    >>> length_of_longest_substring_two_distinct("eceba")
    3
    >>> length_of_longest_substring_two_distinct("ccaabbb")
    5
    """
    # 初始化哈希表，用于记录窗口中每个字符的出现次数
    char_count: Dict[str, int] = {}
    left = 0  # 窗口左边界
    max_len = 0  # 最大子串长度

    # 遍历字符串，使用right作为窗口的右边界
    for right in range(len(s)):
        # 将当前字符加入哈希表，如果字符不存在，则初始化为0再加1
        char_count[s[right]] = char_count.get(s[right], 0) + 1

        # 当窗口内不同字符超过2个时，收缩窗口
        while len(char_count) > 2:
            char_count[s[left]] -= 1  # 减少左侧字符的计数
            if char_count[s[left]] == 0:  # 如果某字符计数为0，则从哈希表中移除
                del char_count[s[left]]
            left += 1  # 移动窗口的左边界

        # 更新最大子串长度
        max_len = max(max_len, right - left + 1)

    return max_len

# 测试代码
test_cases = ["eceba", "ccaabbb"]
results = [length_of_longest_substring_two_distinct(case) for case in test_cases]
print(results)  # [3, 5]